src/commands/sidebar_find_symbol.ts

```diff
@@ -0,0 +1,401 @@
+    this.currentQuery = null;
+  }
+
+  private onDidChangeSelection(selectedElements: (Element)[]) {
+    if (selectedElements.length == 1) {
+      const [element] = selectedElements;
+      if (!(element instanceof Symbol)) {
+        return;
+      }
+      element.show();
+    }
+  }
+
+  /**
+   * Reload the TreeView. This is an alias. Its purpose is to make code prettier.
+   */
+  private reload() {
+    return this.treeView.reload();
+  }
+
+  private setSymbols(lspSymbols: lsp.SymbolInformation[]) {
+    const symbols = lspSymbols.filter((lspSymbol) =>
+      // keep only symbols from real files
+      lspSymbol.location.uri.startsWith("file://")
+    ).map(
+      // turn into `Symbol`s
+      (lspSymbol) => new Symbol(lspSymbol),
+    );
+
+    this.symbols.clear();
+    for (const symbol of symbols) {
+      const { uri } = symbol.location;
+      this.symbols.set(
+        uri,
+        [...(this.symbols.get(uri) ?? []), symbol],
+      );
+    }
+
+    this.files = Array.from(this.symbols.keys()).map((uri) =>
+      new File(uri, this.symbols.get(uri)!)
+    );
+
+    function findDuplicates<Type>(array: Type[]): Type[] {
+      const seenItems: Type[] = [];
+      const duplicates: Type[] = [];
+
+      for (const item of array) {
+        if (seenItems.includes(item)) {
+          duplicates.push(item);
+        }
+        seenItems.push(item);
+      }
+
+      return duplicates;
+    }
+    this.ambiguousFilenames = findDuplicates(
+      this.files.map((file) => file.filename),
+    );
+
+    return this.symbols;
+  }
+
+  displaySymbols(
+    query: string,
+    getSymbols: (query: string) => Promise<lsp.SymbolInformation[] | null>,
+  ) {
+    this.currentQuery = query;
+
+    /**
+     * Create an event listener and remove it after the search query changes.
+     */
+    const updateSymbolsAndManageDisposable = (
+      createEventListener: (callback: () => void) => Disposable,
+    ) => {
+      let disposable: Disposable | null = null;
+      const updateSymbols = async () => {
+        if (this.currentQuery != query) {
+          if (disposable) {
+            // unregister the listener
+            disposable.dispose();
+          }
+          return;
+        }
+
+        const symbols = await getSymbols(query) ?? [];
+        const displayedSymbols = this.setSymbols(symbols);
+
+        if (displayedSymbols.size) {
+          this.headerMessage = `Results for '${query}':`;
+        } else {
+          this.headerMessage = `No results found for '${query}'.`;
+        }
+        this.reload();
+      };
+      updateSymbols();
+
+      disposable = createEventListener(updateSymbols);
+    };
+
+    for (const textEditor of nova.workspace.textEditors) {
+      updateSymbolsAndManageDisposable(
+        textEditor.onDidStopChanging.bind(textEditor),
+      );
+    }
+
+    nova.workspace.onDidAddTextEditor((textEditor) =>
+      updateSymbolsAndManageDisposable(
+        textEditor.onDidStopChanging.bind(textEditor),
```

Comment:
It's very cool that we can update the list of symbols in realtime like this without needing to rerun the query.

But I did notice that the tree view would be rebuilt on every keystroke, which unfortunately triggers an animation, even when there are no changes:

![CleanShot 2022-05-13 at 09 54 37](https://user-images.githubusercontent.com/579491/168237645-29d2767d-70a6-4404-87d4-b5b2f39ab9cd.gif)

This animation is very distracting. Is there a way around it you can think of? Can we only reload if there's a difference? Or is this something we need to ask for in Nova itself?

If it would add a lot of complexity I'm happy to punt and see if we can get Panic to diff tree views on reload, and only animate if there are changes.

Reply:
It's unfortunate; I know. I'll look into this.

Reply:
By the way, I think it's a Nova bug that this is happening on every keystroke. The method we're using is called `onDidStopChanging`, and the documentation says:

> Adds an event listener that invokes the provided callback a short time after the editor stops changing. Multiple changes to text in a short time will be coalesced into one event that can be acted upon performantly.

Reply:
OK —symbol ranges are currently stored in the elements we send to Nova, so we need to reload upon changes for the ranges to be kept accurate. If we didn't reload, clicking on a symbol's name would send the user to the wrong location. 

Instead of storing the ranges in each element object, we could assign an ID to each symbol and keep track of the updated ranges. I'm not sure how easy that would be, though (or whether I'm capable of implementing it at all). If this isn't too great of a problem, I would love for this to be shipped 🚢 🥳 as it works right now. An additional pull request could come later.

On the other hand, look at this Dev. Forum post I made: https://devforum.nova.app/t/callback-sent-to-texteditor-ondidstopchanging-is-called-upon-every-keystroke-more-or-less/1607.

Again, my Dev. Forum username is [unqiue](https://devforum.nova.app/u/unqiue) because I aspire to be unique like Uniqua. (I'm very sorry in case that didn't need to be reiterated. It's just something I'm very passionate about.)

Reply:
This is the [Uniqua you're talking about](https://backyardigans.fandom.com/wiki/Uniqua)? I can respect that.

Let's merge it and open a new issue for the reloading animation. 👍